File: ios/Classes/AnylineTireSidewallPlugin.swift
import Flutter
import UIKit
import AnylineTireTreadSdk

/// Bridges the Tire Sidewall (TSW) scanner to Flutter over the dedicated
/// `anyline_tire_sidewall_plugin` channel.
///
/// Registered from `AnylineTireTreadPlugin.register(with:)` because a Flutter
/// package declares a single `pluginClass` per platform.
public class AnylineTireSidewallPlugin: NSObject, FlutterPlugin {

    private var pendingScanResult: FlutterResult?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: SidewallConstants.CHANNEL,
                                           binaryMessenger: registrar.messenger())
        let instance = AnylineTireSidewallPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case SidewallConstants.METHOD_SCAN:
            scan(result: result, call: call)
        case SidewallConstants.METHOD_IS_SUPPORTED:
            isSupported(result: result)
        case SidewallConstants.METHOD_RESOLVE_PLAY_SERVICES:
            // Android-only; no-op on iOS.
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - SDK Methods

    private func scan(result: @escaping FlutterResult, call: FlutterMethodCall) {
        if pendingScanResult != nil {
            result(failedOutcome(code: .alreadyRunning, message: "A sidewall scan is already in progress"))
            return
        }

        // Present from the topmost view controller (parity with the tread
        // scanner): presenting from the root VC fails when another controller
        // is already presented.
        guard let presenter = topViewController() else {
            result(failedOutcome(code: .invalidArgument,
                                 message: "Unable to present scan: no active view controller"))
            return
        }

        let arguments = call.arguments as? [String: Any]
        let clientId = arguments?[SidewallConstants.EXTRA_CLIENT_ID] as? String ?? ""
        let configJson = arguments?[SidewallConstants.EXTRA_CONFIG_JSON] as? String

        pendingScanResult = result

        AnylineTireSidewallScanner().scan(
            from: presenter,
            clientId: clientId,
            config: buildConfig(configJson)
        ) { [weak self] scanResult in
            DispatchQueue.main.async {
                guard let pending = self?.pendingScanResult else { return }
                self?.pendingScanResult = nil
                pending(self?.serialize(scanResult) ?? [:])
            }
        }
    }

    private func isSupported(result: @escaping FlutterResult) {
        AnylineTireSidewallScanner.companion.isSupported { status, _ in
            DispatchQueue.main.async {
                if let unavailable = status as? TswSupportStatus.Unavailable {
                    result([
                        "supported": false,
                        "userResolvable": unavailable.userResolvable,
                        "error": unavailable.error.toMap(),
                    ])
                } else {
                    // iOS reports the sidewall scanner as always supported.
                    result(["supported": true])
                }
            }
        }
    }

    // MARK: - Helpers

    /// Returns the topmost presented view controller of the key window.
    private func topViewController() -> UIViewController? {
        let windows = UIApplication.shared.connectedScenes
            .compactMap { $0 as? UIWindowScene }
            .flatMap { $0.windows }
        let window = windows.first { $0.isKeyWindow } ?? windows.first
        var top = window?.rootViewController
        while let presented = top?.presentedViewController {
            top = presented
        }
        return top
    }

    private func serialize(_ result: TswScanResult) -> [String: Any] {
        if let completed = result as? TswScanResult.Completed {
            var map: [String: Any] = [
                "kind": "completed",
                "resultJson": completed.resultJson,
                "imageBytes": FlutterStandardTypedData(bytes: toData(completed.imageBytes)),
            ]
            if let lighting = completed.scanMetadata.environmentLighting {
                map["environmentLighting"] = lighting.name
            }
            return map
        }
        if let failed = result as? TswScanResult.Failed {
            return ["kind": "failed", "error": failed.error.toMap()]
        }
        // TswScanResult.Aborted
        return ["kind": "aborted"]
    }

    private func failedOutcome(code: ErrorCode, message: String) -> [String: Any] {
        let error = SdkError(
            code: code,
            type: ErrorType.companion.fromCode(code: code),
            message: message,
            debug: nil
        )
        return ["kind": "failed", "error": error.toMap()]
    }

    private func buildConfig(_ configJson: String?) -> TswScannerConfig {
        let config = TswScannerConfig()
        guard let configJson = configJson,
              let data = configJson.data(using: .utf8),
              let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return config
        }

        if let correlationId = root["correlationId"] as? String {
            config.correlationId = correlationId
        }
        if let texts = root["texts"] as? [String: Any] {
            let t = config.texts
            if let v = texts["initializing"] as? String { t.textInitializing = v }
            if let v = texts["alignTire"] as? String { t.textAlignTire = v }
            if let v = texts["moveCloser"] as? String { t.textMoveCloser = v }
            if let v = texts["moveAway"] as? String { t.textMoveAway = v }
            if let v = texts["faceTire"] as? String { t.textFaceTire = v }
            if let v = texts["ready"] as? String { t.textReady = v }
            if let v = texts["holdSteady"] as? String { t.textHoldSteady = v }
            if let v = texts["focusing"] as? String { t.textFocusing = v }
            if let v = texts["calibratingWhiteBalance"] as? String { t.textCalibratingWhiteBalance = v }
            if let v = texts["calibratingExposure"] as? String { t.textCalibratingExposure = v }
            if let v = texts["tooDark"] as? String { t.textTooDark = v }
        }
        return config
    }

    /// Copies a Kotlin `ByteArray` (bridged as `KotlinByteArray`) into `Data`.
    ///
    /// NOTE: this copies byte-by-byte across the Kotlin/Native boundary. For
    /// large images consider adding a `ByteArray` → `NSData` accessor on the
    /// SDK side to avoid the per-byte bridge cost.
    private func toData(_ array: KotlinByteArray) -> Data {
        let count = Int(array.size)
        guard count > 0 else { return Data() }
        var data = Data(count: count)
        data.withUnsafeMutableBytes { (raw: UnsafeMutableRawBufferPointer) in
            let dst = raw.bindMemory(to: Int8.self).baseAddress!
            for i in 0..<count {
                dst[i] = array.get(index: Int32(i))
            }
        }
        return data
    }
}

File: android/src/main/kotlin/io/anyline/tiretread/flutter/AnylineTireSidewallPlugin.kt
package io.anyline.tiretread.flutter

import android.app.Activity
import android.content.Context
import android.os.Handler
import android.os.Looper
import io.anyline.tiretread.sdk.api.AnylineTireSidewallScanner
import io.anyline.tiretread.sdk.api.ErrorCode
import io.anyline.tiretread.sdk.api.SdkError
import io.anyline.tiretread.sdk.api.TswScanResult
import io.anyline.tiretread.sdk.api.TswSupportStatus
import io.anyline.tiretread.sdk.tsw.ui.configs.TswScannerConfig
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

/**
 * Bridges the Tire Sidewall (TSW) scanner to Flutter over the dedicated
 * `anyline_tire_sidewall_plugin` channel.
 *
 * Registered and driven by [AnylineTireTreadPlugin], which forwards the
 * FlutterPlugin and ActivityAware lifecycle to this handler (a Flutter package
 * declares a single `pluginClass` per platform, so the second channel is set
 * up from the first plugin).
 */
class AnylineTireSidewallPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {

    private lateinit var channel: MethodChannel
    private var applicationContext: Context? = null
    private var activityPluginBinding: ActivityPluginBinding? = null
    private var pendingScanResult: Result? = null

    // The SDK delivers the scan result on a background dispatcher
    // (Dispatchers.Default); MethodChannel replies must be on the main thread.
    private val mainHandler by lazy { Handler(Looper.getMainLooper()) }
    private val mainScope = CoroutineScope(Dispatchers.Main + SupervisorJob())

    private fun onMain(block: () -> Unit) {
        if (Looper.myLooper() == Looper.getMainLooper()) {
            block()
        } else {
            mainHandler.post(block)
        }
    }

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        applicationContext = binding.applicationContext
        channel = MethodChannel(binding.binaryMessenger, SidewallConstants.CHANNEL)
        channel.setMethodCallHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        applicationContext = null
        mainScope.cancel()
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            SidewallConstants.METHOD_SCAN -> {
                val clientId = call.argument<String>(SidewallConstants.EXTRA_CLIENT_ID) ?: ""
                val configJson = call.argument<String>(SidewallConstants.EXTRA_CONFIG_JSON)
                scan(clientId, configJson, result)
            }
            SidewallConstants.METHOD_IS_SUPPORTED -> isSupported(result)
            SidewallConstants.METHOD_RESOLVE_PLAY_SERVICES -> resolvePlayServices(result)
            else -> result.notImplemented()
        }
    }

    // ActivityAware

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activityPluginBinding = binding
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activityPluginBinding = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activityPluginBinding = binding
    }

    override fun onDetachedFromActivity() {
        activityPluginBinding = null
    }

    private fun hostContext(): Context? =
        activityPluginBinding?.activity ?: applicationContext

    // SDK methods

    private fun scan(clientId: String, configJson: String?, result: Result) {
        val context = hostContext()
        if (context == null) {
            result.success(
                failedOutcome(ErrorCode.INVALID_ARGUMENT, "Plugin is not attached to a context"),
            )
            return
        }
        if (pendingScanResult != null) {
            result.success(
                failedOutcome(ErrorCode.ALREADY_RUNNING, "A sidewall scan is already in progress"),
            )
            return
        }

        pendingScanResult = result
        AnylineTireSidewallScanner().scan(
            from = context,
            clientId = clientId,
            config = buildConfig(configJson),
        ) { scanResult ->
            onMain {
                val pending = pendingScanResult ?: return@onMain
                pendingScanResult = null
                pending.success(serialize(scanResult))
            }
        }
    }

    private fun isSupported(result: Result) {
        mainScope.launch {
            val status = runCatching { AnylineTireSidewallScanner.isSupported() }
                .getOrElse { t ->
                    TswSupportStatus.Unavailable(
                        error = SdkError(
                            code = ErrorCode.INTERNAL_ERROR,
                            message = t.message ?: "Failed to check device support",
                        ),
                        userResolvable = false,
                    )
                }
            result.success(serialize(status))
        }
    }

    private fun resolvePlayServices(result: Result) {
        val activity: Activity? = activityPluginBinding?.activity
        if (activity != null) {
            AnylineTireSidewallScanner.resolvePlayServices(activity)
        }
        // No-op when no Activity is available (or already resolvable elsewhere).
        result.success(null)
    }

    // Serialization

    private fun serialize(result: TswScanResult): Map<String, Any?> = when (result) {
        is TswScanResult.Completed -> mapOf(
            "kind" to "completed",
            "resultJson" to result.resultJson,
            // ByteArray is Flutter-codec-safe → surfaces as Uint8List in Dart.
            "imageBytes" to result.imageBytes,
            "environmentLighting" to result.scanMetadata.environmentLighting?.name,
        )
        is TswScanResult.Failed -> mapOf(
            "kind" to "failed",
            "error" to result.error.toMap(),
        )
        TswScanResult.Aborted -> mapOf("kind" to "aborted")
    }

    private fun serialize(status: TswSupportStatus): Map<String, Any?> = when (status) {
        is TswSupportStatus.Unavailable -> mapOf(
            "supported" to false,
            "userResolvable" to status.userResolvable,
            "error" to status.error.toMap(),
        )
        else -> mapOf("supported" to true)
    }

    private fun failedOutcome(code: ErrorCode, message: String): Map<String, Any?> = mapOf(
        "kind" to "failed",
        "error" to SdkError(code = code, message = message).toMap(),
    )

    private fun buildConfig(configJson: String?): TswScannerConfig {
        val config = TswScannerConfig()
        if (configJson.isNullOrBlank()) return config

        val root = runCatching { Json.parseToJsonElement(configJson).jsonObject }
            .getOrNull() ?: return config

        (root["correlationId"]?.jsonPrimitive?.contentOrNull)?.let { config.correlationId = it }

        (root["texts"] as? JsonObject)?.let { texts ->
            fun str(key: String): String? = texts[key]?.jsonPrimitive?.contentOrNull
            str("initializing")?.let { config.texts.textInitializing = it }
            str("alignTire")?.let { config.texts.textAlignTire = it }
            str("moveCloser")?.let { config.texts.textMoveCloser = it }
            str("moveAway")?.let { config.texts.textMoveAway = it }
            str("faceTire")?.let { config.texts.textFaceTire = it }
            str("ready")?.let { config.texts.textReady = it }
            str("holdSteady")?.let { config.texts.textHoldSteady = it }
            str("focusing")?.let { config.texts.textFocusing = it }
            str("calibratingWhiteBalance")?.let { config.texts.textCalibratingWhiteBalance = it }
            str("calibratingExposure")?.let { config.texts.textCalibratingExposure = it }
            str("tooDark")?.let { config.texts.textTooDark = it }
        }
        return config
    }
}
